Write up how third-party auditing / management work (#41)
ietf-wg-keytrans/draft-protocol: draft-ietf-keytrans-protocol.md

original Search or Update query.

# Third Parties

## Management

With the Third-Party Management deployment mode, a third party is responsible
for the majority of the work of storing and operating the Transparency Log,
while the Service Operator serves mainly to enforce access control and
authenticate the addition of new entries. All user queries specified in
{{user-operations}} are initially sent by users directly to the Service Operator
to be forwarded to the Third-Party Manager if they pass access control.

While other operations are forwarded by the Service Operator unchanged,
`UpdateRequest` structures are forwarded to the Third-Party Manager with the
Service Operator's signature attached:

~~~ tls-presentation
struct {
  UpdateRequest request;
  opaque signature<0..2^16-1>;
} ManagerUpdateRequest;
~~~

The signature is computed as described in {{update-format}}. The Service
Operator MUST maintain its own records (independent of the Third-Party Manager)
for the greatest version of each label for the purpose of producing this
signature.

> ## Auditing

Reviewer: I think this section deserves a bit more context as to how auditing is supposed to work out, e.g., say that "auditors will check append-only property at a regular basis," etc.

Author: Background info I think is more appropriate for the arch document. I looked in the arch docu and we're pretty thin on how we describe auditing there too, so I will flesh it out separately

Reviewer: We have a lot of introductory/glue text in other sections already, and I think this is a strength of the draft. Why not just a small intro to third-party management and auditing as intro text to the section "Third Parties"? I think we could be very succinct here, but this would help readers.

With the Third-party Auditing deployment mode, the Service Operator obtains

> signatures from a lightweight Third-Party Auditor attesting to the fact that the

Reviewer: ```suggestion
signatures from a Third-Party Auditor attesting to the fact that the
```

Partially personal preference. "Lightweight" is such an overused term and conveys little meaning. But also: I don't think auditors will be particularly lightweight, no? They may need to perform a lot of computations to check tree construction.

Service Operator is constructing the tree correctly. These signatures are
provided to users along with the responses to their queries.

> Each entry added to the log tree has a corresponding `AuditorUpdate` structure

Reviewer: You say that each entry "has a corresponding `AuditorUpdate` structure. I find it unclear what that means. Will it be added to the log tree itself? I presume not as it is sufficient to show this to single auditors and need not be committed to. Maybe "`AuditorUpdate` structures correspond to log tree entries and ...". Or say that auditors can request them.

that allows the Third-Party Auditor to verify the log entry against stored state
from previous log entries:

~~~ tls-presentation
struct {
  uint64 timestamp;

  PrefixLeaf added<0..2^16-1>;
  PrefixLeaf removed<0..2^16-1>;

  PrefixProof proof;
} AuditorUpdate;

> ~~~

Reviewer: Assuming that these are served to auditors individually, and directly, it may be nice to allow logs to show diffs for more than one log entry, e.g., auditor last checked `n`, and log presents update for `n+m`. This should be easily doable by adding a `uint64 delta_to` field.

Author: I'm not sure what you mean here?

Reviewer: This structure is basically a diff, right? The `timestamp` field encodes the newer version of the diff. My assumption was: Since the structure does not define a field for the older version, the diff is to the previous version (i.e., this structure will always represent a diff from epoch `n` to epoch `n+1`).

I think it would be better to allow for flexibility here and allow auditors to sign off diffs between epochs `n` and `n+m`. For that, we need two timestamps.

Author: AuditorUpdate structures aren't shown to users, they are sent from the service operator to the third-party auditor. The auditor's signature comes back as part of TreeHead, section 9.3. I will push some text that makes this clearer

The `timestamp` field contains the timestamp of the corresponding log entry. The
`added` field contains the list of `PrefixLeaf` structures that were added to
the prefix tree in the corresponding log entry. The `removed` field contains the
list of `PrefixLeaf` structures that were removed from the prefix tree.

The `proof` field contains a batch lookup proof in the previous log entry's
prefix tree for all search keys referenced by `added` or `removed. The
`proof.results` field contains the result of the search for each element of
`added` in the order provided, followed by the result of the search for each
element of `removed` in the order provided.

An auditor processes a single `AuditorUpdate` by following these steps:

1. Verify that `timestamp` is greater than or equal to the timestamp of the
   previous log entry.
2. Verify that the `PrefixSearchResult` provided in `proof` for each element of
   `added` has a `result_type` of `nonInclusionParent` or `nonInclusionLeaf`.
3. Verify that the `PrefixSearchResult` provided in `proof` for each element of
   `removed` has a `result_type` of `inclusion`.
4. For each element of `removed`, verify that, with the addition of the new log

>    entry, the prefix tree leaf will have been published in at least one

Reviewer: Why "at least one"? Why not "the most recent"?

Author: It's possible that if we add a new log entry, that new log entry will be distinguished immediately. This happens for example if the RMW is 0. That would make it the most recent distinguished log entry. But we just removed this search key from it, so it's not in the "most recent" distinguished log entry

>    entry, the prefix tree leaf will have been published in at least one

Reviewer: ```suggestion
   entry, the prefix tree leaf was published in at least one
```
